File: ISS Emergency/Assets/Scripts/TriggerPanel.js
#pragma strict

// Hannes Helmholz
//
// 


private class TriggerPanel extends TriggerAbstract {

	// =============================================================================
	// MEMBERS ---------------------------------------------------------------------
	@SerializeField private var ObjectDescriptionOff:String;
	
	private var ObjectDescriptionOn:String; 
	private var State:boolean = false;
	private var Buttons:TriggerPanelButton[];
	private var Manager:TriggerPanelManager;
	// ============================================================================
	
	
	
	// =============================================================================
	// METHODS UNITY ---------------------------------------------------------------
	
	function Awake() {
		super.Awake();
		
		Buttons = gameObject.GetComponentsInChildren.<TriggerPanelButton>();
		Manager = GameObject.FindObjectOfType(TriggerPanelManager);
		
		ObjectDescriptionOn = ObjectDescription;
	}
	
	
	function OnMouseEnter() {
		if (State)
			NetView.RPC("RPCSetObjectDescription", RPCMode.AllBuffered, ObjectDescriptionOff);
		else
			NetView.RPC("RPCSetObjectDescription", RPCMode.AllBuffered, ObjectDescriptionOn);
		
		super.OnMouseEnter();
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS ---------------------------------------------------------------------
	
	public function SetState(state:boolean) {
		State = state;
			
		super.DoTrigger();
	}
	
	
	protected function DoTrigger() {
		Manager.SetState(!State);
		
		// StatusLast should not be reset on server
		// because focus still on FirePort but Highlighting should be off during measurement
		var last:boolean = StatusLast;
		NetView.RPC("RPCSetStatus", RPCMode.AllBuffered, false);
		StatusLast = last;
	}
	
	
	private function TriggerButtons() {
		for (var button:TriggerPanelButton in Buttons)
			button.Trigger();
	}
	// =============================================================================
	
	
	
	// =============================================================================
	// METHODS RPC -----------------------------------------------------------------
	
	@RPC
	function RPCDoTrigger() {
		super.RPCDoTrigger();
		
		TriggerButtons();
	}
	
	
	@RPC
	function RPCSetObjectDescription(text:String) {
		ObjectDescription = text;
	}
	// =============================================================================
}

File: Assets/Scripts/TriggerPanel.cs

// Hannes Helmholz
//
// 

using UnityEngine;

public class TriggerPanel : TriggerAbstract
{

    // =============================================================================
    // MEMBERS ---------------------------------------------------------------------
    [SerializeField]
    private string ObjectDescriptionOff;
    private string ObjectDescriptionOn;
    private bool State = false;
    private TriggerPanelButton[] Buttons;
    private TriggerPanelManager Manager;
    // ============================================================================
 
 
 
    // =============================================================================
    // METHODS UNITY ---------------------------------------------------------------
 
    protected virtual void Start ()
    {
        base.Start ();
     
        Buttons = GetComponentsInChildren<TriggerPanelButton> ();
        Manager = FindObjectOfType ( typeof( TriggerPanelManager) ) as TriggerPanelManager;
     
        ObjectDescriptionOn = ObjectDescription;
    }
 
 
    protected virtual void OnMouseEnter ()
    {
        if ( State )
        {
            NetView.RPC ( "RPCSetObjectDescription", RPCMode.AllBuffered, ObjectDescriptionOff );
        }
        else
        {
            NetView.RPC ( "RPCSetObjectDescription", RPCMode.AllBuffered, ObjectDescriptionOn );
        }
     
        base.OnMouseEnter ();
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS ---------------------------------------------------------------------
 
    public void SetState ( bool state )
    {
        State = state;
         
        base.DoTrigger ();
    }
 
 
    protected new void DoTrigger ()
    {
        Manager.SetState ( !State );
     
        // StatusLast should not be reset on server
        // because focus still on FirePort but Highlighting should be off during measurement
        bool last = statusLast;
        NetView.RPC ( "RPCSetStatus", RPCMode.AllBuffered, false );
        statusLast = last;
    }
 
 
    private void TriggerButtons ()
    {
        foreach ( TriggerPanelButton button in Buttons )
            button.Trigger ();
    }
    // =============================================================================
 
 
 
    // =============================================================================
    // METHODS RPC -----------------------------------------------------------------
 
    [RPC]
    void RPCDoTrigger ()
    {
        base.RPCDoTrigger ();
     
        TriggerButtons ();
    }
 
 
    [RPC]
    void RPCSetObjectDescription ( string text )
    {
        ObjectDescription = text;
    }
    // =============================================================================
}